include/vulkan/utility/vk_typemap_helper.hpp

```diff
@@ -0,0 +1,1890 @@
+// returns nullptr if the entry is not found
+template <typename T> const T *FindStructInPNextChain(const void *next) {
+    const VkBaseOutStructure *current = reinterpret_cast<const VkBaseOutStructure *>(next);
+    VkStructureType desired_sType = GetSType<T>();
+    while (current) {
+        if (desired_sType == current->sType) {
+            return reinterpret_cast<const T*>(current);
+        }
+        current = current->pNext;
+    }
+    return nullptr;
+}
+
+// Find an entry of the given type in the non-const pNext chain
+// returns nullptr if the entry is not found
+template <typename T> T *FindStructInPNextChain(void *next) {
+    VkBaseOutStructure *current = reinterpret_cast<VkBaseOutStructure *>(next);
+    VkStructureType desired_sType = GetSType<T>();
+    while (current) {
+        if (desired_sType == current->sType) {
+            return reinterpret_cast<T*>(current);
+        }
+        current = current->pNext;
+    }
+    return nullptr;
+}
+
+// Find last element of pNext chain
+inline VkBaseOutStructure *FindLastStructInPNextChain(void *next) {
+    auto *current = reinterpret_cast<VkBaseOutStructure *>(next);
+    auto *prev = current;
+    while (current) {
+        prev = current;
+        current = reinterpret_cast<VkBaseOutStructure *>(current->pNext);
+    }
+    return prev;
+}
+
+// Init the header of an sType struct
+template <typename T>
+T InitStruct(void *p_next = nullptr) {
+    T out = {};
+    out.sType = GetSType<T>();
+    out.pNext = p_next;
+    return out;
+}
+
+// Init the header of an sType struct with pNext and optional fields
+template <typename T, typename... StructFields>
+T InitStruct(void *p_next, StructFields... fields) {
+    T out = {GetSType<T>(), p_next, fields...};
+    return out;
+}
+
+class InitStructHelper {
+    void* p_next = nullptr;
+  public:
+    InitStructHelper() = default;
+    InitStructHelper(void *p_next) : p_next(p_next) {};
+
+    template <typename T>
+    operator T() {
+        return InitStruct<T>(p_next);
+    }
+};
+
+} // namespace vku
+
+// Init the header of an sType struct with pNext and optional fields
+template <typename T, typename... StructFields>
+T LvlInitStruct(void *p_next, StructFields... fields) {
+    T out = {vku::GetSType<T>(), p_next, fields...};
+    return out;
+}
+
+// Init the header of an sType struct
+template <typename T>
+T LvlInitStruct(void *p_next = nullptr) {
+    T out = {};
+    out.sType = vku::GetSType<T>();
+    out.pNext = p_next;
+    return out;
+}
+
+// Find an entry of the given type in the const pNext chain
+template <typename T> const T *LvlFindInChain(const void *next) { return vku::FindStructInPNextChain<T>(next); }
+
+// Find an entry of the given type in the pNext chain
+template <typename T> T *LvlFindModInChain(void *next) { return vku::FindStructInPNextChain<T>(next); }
+
+// Find last element of pNext chain
+inline VkBaseOutStructure *LvlFindLastInChain(void *next) { return vku::FindLastStructInPNextChain(next); }
```

Comment:
👎🏾 

If we don't fix this now when will we fix it? I understand it will be a pain to fix.

Reply:
Thats fair, why have long term pain when some short term will lessen it.